FernFlower-Patches/0048-Reduce-allocations-in-getAllExprents.patch

```diff
@@ -0,0 +1,269 @@
+index 6ed64f77930f44a9c6bd17d2eca51de97a6cbbcf..805e0ee518123d6578abcd60d2c297d2f437bec0 100644
+--- a/src/org/jetbrains/java/decompiler/modules/decompiler/exps/ConstExprent.java
++++ b/src/org/jetbrains/java/decompiler/modules/decompiler/exps/ConstExprent.java
+@@ -164,8 +164,8 @@ public class ConstExprent extends Exprent {
+   }
+ 
+   @Override
+-  public List<Exprent> getAllExprents() {
+-    return new ArrayList<>();
++  public List<Exprent> getAllExprents(List<Exprent> list) {
++    return list;
+   }
+ 
+   @Override
+diff --git a/src/org/jetbrains/java/decompiler/modules/decompiler/exps/ExitExprent.java b/src/org/jetbrains/java/decompiler/modules/decompiler/exps/ExitExprent.java
+index 996eb3c3fe5388f7fde11df8937a09d01843f89d..0de6ad2cfa8f7dde5b9142a10525b58853f0549a 100644
+--- a/src/org/jetbrains/java/decompiler/modules/decompiler/exps/ExitExprent.java
++++ b/src/org/jetbrains/java/decompiler/modules/decompiler/exps/ExitExprent.java
+@@ -59,8 +59,7 @@ public class ExitExprent extends Exprent {
+   }
+ 
+   @Override
+-  public List<Exprent> getAllExprents() {
+-    List<Exprent> lst = new ArrayList<>();
++  public List<Exprent> getAllExprents(List<Exprent> lst) {
+     if (value != null) {
+       lst.add(value);
+     }
+diff --git a/src/org/jetbrains/java/decompiler/modules/decompiler/exps/Exprent.java b/src/org/jetbrains/java/decompiler/modules/decompiler/exps/Exprent.java
+index 8074782862f3842a95050212df0ac5d778f20091..4a696f53b52d3a577b7a1258c37e40996952b576 100644
+--- a/src/org/jetbrains/java/decompiler/modules/decompiler/exps/Exprent.java
++++ b/src/org/jetbrains/java/decompiler/modules/decompiler/exps/Exprent.java
+@@ -99,14 +99,25 @@ public abstract class Exprent implements IMatchable {
+     return false;
+   }
+ 
+-  public List<Exprent> getAllExprents(boolean recursive) {
+-    List<Exprent> lst = getAllExprents();
++  public final List<Exprent> getAllExprents(boolean recursive) {
++    List<Exprent> lst = new ArrayList<>();
++    getAllExprents(recursive, lst);
++
++    return lst;
++  }
++
++  private List<Exprent> getAllExprents(boolean recursive, List<Exprent> list) {
++    int start = list.size();
++    getAllExprents(list);
++    int end = list.size();
++
+     if (recursive) {
+-      for (int i = lst.size() - 1; i >= 0; i--) {
+-        lst.addAll(lst.get(i).getAllExprents(true));
++      for (int i = end - 1; i >= start; i--) {
++        list.get(i).getAllExprents(true, list);
+       }
+     }
+-    return lst;
++
++    return list;
+   }
+ 
+   public Set<VarVersionPair> getAllVariables() {
+@@ -122,10 +133,17 @@ public abstract class Exprent implements IMatchable {
+     return set;
+   }
+ 
+-  public List<Exprent> getAllExprents() {
+-    throw new RuntimeException("not implemented");
++  public final List<Exprent> getAllExprents() {
++    List<Exprent> list = new ArrayList<>();
++    getAllExprents(list);
++
++    return list;
```

Comment:
Wouldn't this make slightly more sense as `return getAllExprents(list);` or even `return getAllExprents(new ArrayList<>());`